### PR TITLE
WIP: Telegram Passport support

### PR DESCRIPTION
**Pull request by tjhorner**
# Description

**This pull request is a work in progress.** Still needed:

- Unit tests
- `getPassportFile` method for retrieving + decrypting Passport files all at once

This pull request adds support for Telegram Passport and decryption of Telegram Passport data with the [`telegram-passport`](https://www.npmjs.com/package/telegram-passport) package.

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

A simple way to test this change would be:

```js
const Telegraf = require('./telegraf.js')

const bot = new Telegraf("BOT_TOKEN", {
  privateKey: "BOT_PRIVATE_KEY_IN_PEM_FORMAT"
})

bot.on("passport_data", ctx => {
  console.log(ctx.decryptPassportData()) // will print the decrypted Passport data
})

bot.startPolling()
```

**Test Configuration**:
* Node.js Version: v8.4.0
* Operating System: Windows 10 Home / Ubuntu 16.04

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/context.js [hunk 1]
```diff
@@ -101,6 +103,10 @@ class TelegrafContext {
     return this.update.callback_query
   }
 
+  get passportData () {
+    return this.update.passport_data
```
Comment (dotcypress) [inline core/context.js]: `passport_data` is a part of `Message`, not `Update`:

`return this.message && this.message.passport_data`

File: core/context.js [hunk 2]
```diff
@@ -146,6 +152,15 @@ class TelegrafContext {
     }
   }
 
+  decryptPassportData () {
+    if (this.options.privateKey) {
+      var passport = new TelegramPassport(this.options.privateKey)
+      return passport.decrypt(this.update.message.passport_data)
```
Comment (dotcypress) [inline core/context.js]: Why not `return passport.decrypt(this.passportData)`?

File: core/context.js (hunk 1 above)
Comment (tjhorner) [reply core/context.js]: Good catch, thanks!

File: core/context.js (hunk 2 above)
Comment (tjhorner) [reply core/context.js]: I think I wrote this before I added the `passportData` property. I'll update this in the next commit

**Comment by tjhorner** [comment]
Do we want something that automatically fetches and then decrypts files, maybe an extension to `getFile`. `getPassportFile`?

**Comment by dotcypress** [comment]
Yes, I think `getPassportFile` will be useful in some cases! 👍

**Comment by Piterden** [comment]
Btw guys, it's not fully related, but could be interesting for you https://github.com/sergeysova/telegram-typings/issues/23

**Comment by lifehome** [comment]
Hi, sorry for bumping this but will this feature be updated? I am looking for Passport support and this seem looks good.

**Comment by lifehome** [comment]
Will there be chances this issue get resolved in 2019? :wonder:

**Comment by tjhorner** [comment]
@lifehome the only thing blocking this PR from being merged is tests. Unfortunately, I don't have the time to write them myself, but if someone else wants to contribute I would be glad to accept a PR into my own branch!

**Comment by dotcypress** [comment]
@tjhorner thank you a lot for `telegram-passport` library and this PR 👍🔥. 
I decided to use delegation pattern with passport feature instead of adding new dependency:

```js
const Telegraf = require('telegraf')
const TelegramPassport = require('telegram-passport')

const bot = new Telegraf(process.env.BOT_TOKEN)
const passport = new TelegramPassport("PRIVATE_KEY_IN_PEM_FORMAT")

bot.on('passport_data', (ctx) => {
  const decryptedPasswordData = passport.decrypt(ctx.passportData)
  console.log(decryptedPasswordData)
  return ctx.setPassportDataErrors([
    { source: 'selfie', type: 'driver_license', file_hash: 'file-hash', message: 'Selfie photo is too low quality'}
  ])
})

```

New version will be published shortly.

**Pull request closed without merging**